### DISCUSSION TITLE
Receive SYSEX messages via MIDI input

### DISCUSSION THREAD

**Discussion by jrdooley** [probonopd/MiniDexed#110] [Q&A] [answered]
Does MiniDexed currently support receiving SYSEX messages via a MIDI connected device to allow the real-time changing of parameters?

For example, by connecting a class-compliant USB MIDI interface to the RPi, SYSEX messages could be sent to RPi/MiniDexed from an external device (e.g. a computer with sequencer sending SYSEX messages) updating OP1 Freq Coarse parameter immediately. This would allow for an external interface to program voices.

This question is based on the presupposition that rather than connecting a USB MIDI keyboard, a class-compliant MIDI interface can be used instead. Is this possible?

**Comment by probonopd** [comment, marked as answer]
Hello @jrdooley. 

__Update:__ This is now implemented thanks to @dcoredump.

> This question is based on the presupposition that rather than connecting a USB MIDI keyboard, a class-compliant MIDI interface can be used instead. Is this possible?

Yes.

**Reply by jrdooley** [reply]
@probonopd Great to hear that a MIDI interface can be used. +1 for ticket #82 being implemented!

**Comment by miotislucifugis** [comment]
It looks like the pull request for handling sysex parameter changes has been merged, but im not having any luck getting this to work.  Trying to used dexed to change sound setting while playing. Algorithm, etc.     Should this be working or is it still "in progress?"   Midi is definitely being sent and received by miniDexed, as i can trigger notes on miniDexed with the Dexed app keyboard.  using the most recent continuous build version (from today)

**Reply by dcoredump** [reply]
As far as I have read, it should work. Probably Dexed does not seem to send SYSEX. Please try [EdiSyn](https://github.com/eclab/edisyn).

**Reply by dcoredump** [reply]
> As far as I have read, it should work. Probably Dexed does not seem to send SYSEX. Please try [EdiSyn](https://github.com/eclab/edisyn).

Or try [Herr Müller Online](https://dx7.vstforx.de/) - my favorite Editor.

**Reply by miotislucifugis** [reply]
Dexed definitely does send sysex, as I've used it many times before back when I had a DX7 and a tx7, both for patch/cart dumps and tweaking individual parameters while playing.     Also, i can see the sysex messages being sent in midiMonitor.  I'll give it a go with Edisyn later and report back.   I dont need to enable anything in minidexed.ini, do I?

**Comment by sou7611** [comment]
What method are you guys using for MIDI on the Pi ... connecting directly
to GPIO pins with interface circuitry, or plugging in a USB MIDI ??? and if
latter, which one?

**Comment by luisfcorreia** [comment]
Take a look here:
https://github.com/probonopd/MiniDexed/wiki/Hardware#classic-midi-input

**Comment by sou7611** [comment]
I read that ... what I'm wondering is if a single port USB midi interface
qualifies, or if the only way to plug a midi cable in is to build the
interface.

Not that I'm adverse to that, but I already have the former ....

**Reply by craigyjp** [reply]
It would have to be supported by circle, I guess if there is one then it should work.

**Reply by miotislucifugis** [reply]
If im not mistaken, currently you must build a serial Midi In circuit.

**Comment by luisfcorreia** [comment]
Let me ask you:

What device do you intend to use with minidexed?

**Comment by miotislucifugis** [comment]
@dcoredump ,
still not having any luck sending sysex parameters.   Using the latest build (after double note fix), and EdiSyn.   For the most part, trying to change Algorithm.   When I turn the "knob"  I can see that sysex is being sent via midi monitor and a light on my interface.   in Edisyn, if I click on midi - send to current patch, there is stuff received, as the sound changes, but it changes everything to whatever the default edisyn patch setting are.

**Comment by dcoredump** [comment]
>  @dcoredump ,
> still not having any luck sending sysex parameters.   Using the latest build (after double note fix), and EdiSyn.   For the most part, trying to change Algorithm.   When I turn the "knob"  I can see that sysex is being sent via midi monitor and a light on my interface.   in Edisyn, if I click on midi - send to current patch, there is stuff received, as the sound changes, but it changes everything to whatever the default edisyn patch setting are.  

Thanks for reporting!

I hope I can fix my MiniDexed soon, so that I can debug and hopefully fix this.

**Reply by miotislucifugis** [reply]
no worries, no rush...   Just wanted to let you know and check if things were working properly on my end.   By the way, while edisyn seems nice and versatile, if you arent familiar with it already, I highly recommend checking out Dexed.  either vst or the standalone.  It may be helpful for testing, as all the parameters are laid out in one page and it will be prob the defacto interface for most users managing/editing patches on minidexed.   cheers

**Comment by dcoredump** [comment]
>  @dcoredump ,
> still not having any luck sending sysex parameters.   Using the latest build (after double note fix), and EdiSyn.   For the most part, trying to change Algorithm.   When I turn the "knob"  I can see that sysex is being sent via midi monitor and a light on my interface.   in Edisyn, if I click on midi - send to current patch, there is stuff received, as the sound changes, but it changes everything to whatever the default edisyn patch setting are.  

I made a quick test: with my setup I can see MIDI events on my scope and MiniDexed recognizes them. I could not test if anything changes with the sound (will do this later).

![PXL_20220515_133134415.jpg](https://user-images.githubusercontent.com/5188678/168475577-dca579f4-f4b7-4fa8-9c92-0f02f211cb4c.jpg)

Can you describe your setup? Are you using DIN-MIDI-Input? That's currently the only way to send SYSEX if you don't have a keyboard which can send it.

[EDIT:] Maybe you have to receive the SYSEX changes on MIDI channel 1. It's a little bit tricky with SYSEX. Normally SYSEX messages don't have a MIDI channel information. But SYSEX for a DX7 seems to have a "device identifier". I implemented this identifier as the MIDI-channel. But I don't know what Dexed or other editor software uses. Perhaps they use device number 1... what I use as MIDI-channel 1...

**Reply by probonopd** [reply]
Very nice oscilloscope @dcoredump :+1:

**Reply by probonopd** [reply]
By the way, it took me a while to realize, but Dexed on the PC can be used just like EdiSyn.

Parm -> DX7 Out -> You have to choose your USB-MIDI-to-serial device that is connected to MiniDexed here
Options -> Audio/MIDI Settings -> MIDI Output -> You have to choose your USB-MIDI-to-serial device that is connected to MiniDexed here, too

Then when you turn a knob in Dexed the respective Sysex are sent to (and working on) MiniDexed. The only thing that is not working (yet?) is sending whole voices or cartridges.

**Reply by dcoredump** [reply]
> By the way, it took me a while to realize, but Dexed on the PC can be used just like EdiSyn.
> 
> Parm -> DX7 Out -> You have to choose your USB-MIDI-to-serial device that is connected to MiniDexed here
> Options -> Audio/MIDI Settings -> MIDI Output -> You have to choose your USB-MIDI-to-serial device that is connected to MiniDexed here, too
> 
> Then when you turn a knob in Dexed the respective Sysex are sent to (and working on) MiniDexed. The only thing that is not working (yet?) is sending whole voices or cartridges.

Will check this soon and fix it.

**Reply by dcoredump** [reply]
> Very nice oscilloscope @dcoredump :+1: 

Borrowed from a friend. Completely oversized, unless you want to build FM transmitters.

And: it's Windows based 🙊

**Comment by miotislucifugis** [comment]
Strange... sending whole patch dumps is the only thing that is working with mine...with edisyn at least.

Im using both Dexed and Edisyn on my PC.  I also have midiMonitor running which displays all midi info on all the busses- the computer is connected to a MOTU audio interface which has dedicated DIN Midi In and Out jacks.  Out is connected to the minidexed, which has a reliable serial midi in ciruit (seemingly reliable at least-  no problem with non-sysex messages.)  All settings in dexed and edisyn are for OUT on the motu on ch#1.
As I mentioned above, MiniDexed does respond to sysex, but only if i click "send to current patch" in Edisyn... this sends the settings for the entire control panel.  When I tweak an individual parameter on either software, i can see a shorter sysex message (ex: F0 10 43 41 01 06 0B 7F) in midimonitor and see the Midi out icon on the LCD of the Motu blink, but the minidexed doesnt respond.   It only seems to like the longer ones that have all the settings ie patch dumps.

**Reply by probonopd** [reply]
Turns out that **voice loading** (using Dexed on the PC) is __working__ but for it to work **one must not set the tone generator(s) on MiniDexed to MIDI channel "Omni"** but to channel 1 (or another channel that is set in Dexed on the PC under PARM -> DX7 Channel).

**Live voice editing** (using Dexed on the PC) is __also working__ but for it to work **one MUST set the tone generator(s) on MiniDexed to MIDI channel "Omni"**.

What gives? Bug or intentional?

**Comment by sou7611** [comment]
I'd hazard a guess at intentional, as you would generally want sysex tied to a specific midi channel.

On May 15, 2022 7:41:13 PM EDT, probonopd ***@***.***> wrote:
>Turns out that **voice loading** (using Dexed on the PC) is __working__ but for it to work **one must not set the tone generator(s) on MiniDexed to MIDI channel "Omni"** but to channel 1 (or another channel that is set in Dexed on the PC under PARM -> DX7 Channel).
>
>**Live voice editing** (using Dexed on the PC) is __also working__ but for it to work **one MUST set the tone generator(s) on MiniDexed to MIDI channel "Omni"**.
>
>What gives? Bug or intentional?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/probonopd/MiniDexed/discussions/110#discussioncomment-2755832
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by miotislucifugis** [comment]
Good catch.  I just set a TG to omni and receiving individual parameter changes is working.  Great.      
 I do not recall that being the behavior when I had a DX7 and TX7

**Reply by probonopd** [reply]
Maybe @dcoredump knows?

**Reply by dcoredump** [reply]
> Maybe @dcoredump knows?

I have to check the code, but maybe the problem is the same as I wrote [here](https://github.com/probonopd/MiniDexed/discussions/110#discussioncomment-2753997).

I thought about a minidexed.ini parameter like `EnableMIDIChannelAsSysExDevice=true`. So the Device ID is the dame as the MIDI channel. If set to `no` each TG must have a `SysExDeviceID`.

But first all errors have to be removed...

**Reply by probonopd** [reply]
As a start, let's try to match the behavior of TX816/TX802. So that we can use existing hardware/software editors out there.

**Reply by dcoredump** [reply]
I have found a problem with the SysEx MIDI channel code but cannot create a PR due to merge problems from upstream. Have to fix this first.

**Comment by sou7611** [comment]
Something similar-ish to this

TENINYU USB to MIDI Cable Converter 2 in 1 PC to Synthesizer Music Studio
Keyboard Interface Wire Plug Controller Adapter Cord 16 Channels Supports
Computer Laptop Windows and Mac
https://www.amazon.com/dp/B07L8KFYBK/ref=cm_sw_r_apan_i_5J6R2WBGP6DED0YR790Z

**Reply by diyelectromusic** [reply]
Those kinds of devices ought to work fine, but be warned, the cheap ones can be very problematic in practice... I tend to only use my Roland UM-ONE these days for example.

More here: http://www.arvydas.co.uk/2013/07/cheap-usb-midi-cable-some-self-assembly-may-be-required/

Kevin

**Reply by maxcomperatore** [reply]
Quick warning for anyone using those generic USB-to-MIDI cables (like TENINYU or the translucent round 3-LED converters) with SysEx:

These cheap cables work fine for basic 3-byte Note-On/Note-Off and CC messages, but almost all of them fail on SysEx bulk dumps. 

The hardware inside uses cheap cloned microcontrollers with small USB endpoint buffers (often 64 bytes or less) and omitted optocouplers on the DIN side. When you fire a 155-byte VCED edit buffer dump or a 4,104-byte 32-voice bank, the cable buffer overflows and drops bytes mid-stream, resulting in corrupt patches or checksum errors.

If you are transferring voice dumps or editing parameters via SysEx, sticking to class-compliant interfaces with proper buffer handling (like Roland UM-ONE mk2, Yamaha UX16, or audio interfaces with DIN ports) will save you hours of debugging.

We had to document this exact cable issue for [bipluk](https://bipluk.com) users because 90% of "failed sysex transfer" reports in the browser turn out to be unbuffered $5 USB cables dropping packets.

**Reply by penfold42** [reply]
> Quick warning for anyone using those generic USB-to-MIDI cables (like TENINYU or the translucent round 3-LED converters) with SysEx:
[..]
> We had to document this exact cable issue for [bipluk](https://bipluk.com) users because 90% of "failed sysex transfer" reports in the browser turn out to be unbuffered $5 USB cables dropping packets.

They are however a good cheap source of case and cables to squeeze an arduino pro micro into!

**Reply by maxcomperatore** [reply]
Haha, 100%! Hollow out the cloned chip, throw in a Pro Micro with LUFA / MIDI firmware and an optocoupler on the RX pin, and suddenly you have a rock-solid USB-MIDI adapter in a factory molded case.

**Reply by probonopd** [reply]
Can you recommend exact firmware and circuit? I need to build some...

**Reply by penfold42** [reply]
> Cab you recommend exact firmware and circuit? I need to build some...

My current cable used https://github.com/tttapa/Control-Surface
BUT! it has an issue - If  I plug it into a synth that's sending clock and active sensing AND I dont have anything consuming that data, you can't send.
This is with ubuntu. I suspect it has a shared buffer for TX and RX thats getting full and blocking.

It's on my long list of projects to get back to :)

**Reply by maxcomperatore** [reply]
For building an ultra-reliable USB-to-MIDI cable that handles SysEx bulk dumps without breaking a sweat:

### 1. Hardware Schematic (Pro Micro ATmega32U4)
The classic culprit with generic cables is the lack of a proper optocoupler and unbuffered UART queues.

```text
[DIN MIDI IN]
Pin 4 (Current Source) ---> 220R Resistor ---> Anode (Pin 2) of 6N138 / H11L1
Pin 5 (Current Return) ---> Cathode (Pin 3) of 6N138
                             (1N4148 diode reverse-parallel across pins 2 & 3 for back-EMF protection)
Pin 2 (Ground) ------------> Unconnected on MIDI IN (prevents ground loops)

[Optocoupler 6N138 Output Side]
Pin 8 (VCC) ---------------> +5V (VCC on Pro Micro)
Pin 5 (GND) ---------------> GND on Pro Micro
Pin 6 (VO Output) ---------> RX1 (PD2 / Pin 0 on Pro Micro) with a 1k pull-up resistor to +5V
Pin 7 (Base) --------------> 4.7k resistor to GND (cleans up optocoupler rise time for 31.25k baud)

[DIN MIDI OUT]
Pin 4 ---------------------> 220R Resistor ---> +5V (VCC)
Pin 5 ---------------------> 220R Resistor ---> TX1 (PD3 / Pin 1 on Pro Micro)
Pin 2 ---------------------> GND
```

### 2. Firmware Options (Handling Active Sensing & Large SysEx)
Regarding the issue @penfold42 ran into where Clock and Active Sensing (`0xFE`) congest the buffer:

1. **`lufa-lib` Dual-Buffer MIDI** (Most robust):
   Using Dean Camera's LUFA MIDI Class Driver on the ATmega32U4 with interrupt-driven ring buffers (at least 256 bytes) avoids the blocking USB endpoint halt.
2. **Filtering Real-Time Bytes**:
   If the connected synth constantly spams Active Sensing (`0xFE`) or MIDI Clock (`0xF8`) and your downstream listener isn't draining them, you can drop or bypass `0xFE` in the hardware UART interrupt ISR before it ever enters the USB endpoint FIFO.
3. **Alternative Ready-to-Flash**:
   - [Midi-USB-Interface by thecalcaholic](https://github.com/thecalcaholic/Midi-USB-Interface): Clean, dedicated USB-MIDI class-compliant firmware for ATmega32U4 with configurable SysEx and real-time filtering.
   - [usb-midi-cloned-cable-fix by FortySevenEffects](https://github.com/FortySevenEffects/arduino_midi_library): Strips the cheap glop-top PCB from the transparent Amazon cable housing and fits the Pro Micro directly into the original shell.
